contributors/design-proposals/storage/csi-migration.md

```diff
@@ -661,6 +729,117 @@ with minimal disruption to users would be:
+have plugin-specific migration feature flag enabled) in appendAttachableLegacyProviderVolumes()
+and appendLegacyProviderVolumes() in the Kubernetes Controller Manager and Kubelet.
+All functions in Kubernetes code base that depend on probed in-tree plugins need to be
+audited and refactored to handle errors returned (due to absence of a probed plugin).
+
+### Enhancements in Probing/Registration of in-tree plugins
+
+Functions appendLegacyProviderVolumes (in the Kubernetes Controller Manager and Kubelet)
+and appendAttachableLegacyProviderVolumes (in the Kubernetes Controller Manager)
+will be enhanced to [1] check plugin specific migration feature flags as well as
+[2] the overall CSIMigrationInTreeOff feature flag to determine whether ProbeVolumePlugins
+function of a legacy in-tree plugin will get invoked.
+
+Once CSIMigrationInTreeOff feature flag and the plugin specific migration
+flags get enabled by default, the build tag `nolegacyproviders` can be enabled
+for testing purposes.
+
+### Detection of migration status for a plugin
+
+Code paths that need to check migration status of a plugin, for example,
+in provisionClaimOperationExternal, findDeletablePlugin, etc. will need to
+depend on an instance of the CSIMigratedPluginManager that provides the following
+utilities:
+```
+func (pm CSIMigratedPluginManager) IsCSIMigrationEnabledForPluginByName(pluginName string) bool
+func (pm CSIMigratedPluginManager) IsPluginMigratableToCSIBySpec(spec *Spec) (bool, error)
+```
+The CSIMigratedPluginManager will be introduced in pkg/volume/csi_migration.go
+
+Note that a per-plugin member function of the form IsMigratedToCSI cannot be used
+since [1] a plugin object for in-tree plugins will typically be nil and [2] the
+code implementing IsMigratedToCSI will be removed as part of disabling plugin code.
+
+### Handling of errors returned by FindPluginBySpec/FindPluginByName for legacy plugins
+
+Once an in-tree plugin is no longer probed (through ProbeVolumePlugins), all the VolumePluginMgr
+functions of the form Find*PluginBySpec/Find*PluginByName will return an error.
+For example, invocation of:
+
+1. FindProvisionablePluginByName in the pv controller will return error for a migrated
+plugin that is no longer probed in ProbeControllerVolumePlugins.
+2. FindExpandablePluginBySpec in the expand controller will return error for a migrated
+plugin that is no longer probed in ProbeExpandableVolumePlugins.
+3. FindAttachablePluginBySpec in the attach/detach controller will return error for
+a migrated plugin that is no longer probed in ProbeAttachableVolumePlugins.
+
+Code invoking the above functions will need to either:
+1. Check for migration status of a plugin based on Volume spec or Plugin name before
+invoking the above functions so that an error is never returned due to missing plugins.
+2. Handle an error from the above functions resulting from missing plugins appropriately.
+This should be rare. 
```

Comment:
I'm still not convinced this should ever happen. But might be an implementation detail - we can leave the design as is for now

Reply:
It is indeed an implementation detail and is only relevant for FindProvisionablePluginByName. Will mention this explicitly.